### PR TITLE
Adding softmax, cross-entropy and cost module

### PR DESCRIPTION
**Pull request by afshinm**

### CODE REVIEW

**Comment by scriptandcompile** [comment]
So the point of the enum from name of the costfunction is to figure out which cost function to apply? I can't help but think we are producing a hand rolled version of polymorphism here. I'm sure there is a better way to do this, perhaps some macro magic I'm ignorant of.

**Comment by afshinm** [comment]
Yes, that's the reason for having that Enum. I need to match the
CostFunction trait and make a decision in NeuralNetwork.

**Comment by afshinm** [comment]
I just added a TODO to address that issue later. Feel free to send a PR / add any comments. Thanks!
